lib/kubernetes-deploy/kubernetes_resource/persistent_volume_claim.rb

```diff
@@ -3,12 +3,66 @@ module KubernetesDeploy
   class PersistentVolumeClaim < KubernetesResource
     TIMEOUT = 5.minutes
 
+    def init_storage_class(cache)
+      @storage_class = {}
+
+      if @definition.dig("spec", "storageClassName").nil?
+        # if no storage class is defined we try to find the default one
+        # FIXME: This assumes the DefaultStorageClass admission plugin is turned on,
+        # need a way to determine this
+        is_default_class = "storageclass.beta.kubernetes.io/is-default-class"
+
+        default_sc = cache.get_all("StorageClass").select do |sc|
+          sc.dig("metadata", "annotations", is_default_class) == "true"
+        end
+
+        if default_sc.length != 1
+          warn_msg = "Multiple default StorageClasses found. If the DefaultStorageClass " \
+            "admission plugin is turned on, all PVC creation will fail."
+          @logger.warn(warn_msg) if default_sc.length > 1
+          return
+        else
+          # using default storage class
+          sc_name = default_sc[0]["metadata"]["name"]
+        end
+      else
+        # using storage class from pvc definition
+        sc_name = @definition.dig("spec", "storageClassName")
+
+        # a storageClassName of "" is an explicit way of saying you want a PV
+        # with no defined StorageClass. We won't look up a storage_class
+        # if that is the case
+        return if sc_name == ""
+      end
+
+      @storage_class = cache.get_instance("StorageClass", sc_name)
+
+      # check the defined StorageClass exists
+      warn_msg = "StorageClass/#{sc_name} not found. This is required for #{id} to deploy."
```

Comment:
Re: StorageClass - I wasn't sure what your preference would be here (since StorageClass is not namespaced). But if the StorageClass is present in the deployment, deploying it first would resolve this issue. A change here probably deserves its own PR though.

Re: the warn message - I did want to place it earlier. The reason it's here is because I needed the cache object to examine the StorageClass (and I don't think I can get it earlier?).

In terms of severity - again I wasn't sure what your preference would be. I used `kubectl apply` as my guide here. It just deploys the PVC and leaves it in the PENDING state. But obviously its role is different to `kubernetes-deploy`s.

But I'm happy to change that `warn` to a failure. Would I do that with something similar to `raise_predates_deploy_error` in `pod.rb`?

Reply:
I think the right thing to do here is set `storage_class` back to nil since it could be created during the deploy. Then add override the default `timeout_message` with this error.

Reply:
I think my initial understanding of how this works was incorrect. In doing more testing, I've found that not deploying the StorageClass at all can still allow success. It appears a default StorageClass definition is being used that has no provisioner and a 'volumeBindingMode: Immediate' (I should check the k8s code to confirm).

As a result, I have found that as long as the PV is present, the PVC will bind to it immediately and the deployment will succeed.

So I think the `timeout_message` should suggest that Dynamic Volume Provisioning may have failed because the StorageClass is not defined. But even if it succeeds - it may have used a volumeBindingMode of Immediate when then intention was to use WaitForFirstConsumer. Should I add a WARN for that somewhere? For example: "StorageClass not found; volumeBindingMode: Immediate will be used"

Reply:
In ebb89e78026b64a19b3cccf3a4e78e3d2d4d997d I replaced the warn with a timeout_msg